### PR TITLE
Mapping exploration

### PR DESCRIPTION
**Pull request by BarceloChristian**
# Description

Cartographer has been removed from packages.txt and installation of the latest version is being made in the docker.
Fixed version of explore-lite (a.k.a m-explore) has been added to .repos.
A new package called mapping_exploration has been added to generate launchfiles to initialize exploration nodes.
A launchfile for explore-lite has been added.

## Explore-lite configuration:
- Base frame and move base topic changed.
- Min frontier size increased a little to 0.8m
- Transform tolerance increased to 0.5m
- The potential scale (weighs frontiers based on distance) has not been changed.
- The gain scale (weighs frontiers based on size) has been reduced to 0. In this way, the robot will move more fluent toward it closest frontier.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Run any kind of simulation using the proper configuration for cartographer
```
export RVIZ=true
export LOCALIZATION=cartographer
export LASER=rplidar
```
Then launch a simulation, e.g:
```
roslaunch ca_gazebo create_house.launch
```
And following that, as soon as the simulation starts:
```
roslaunch ca_mapping_exploration explore_lite.launch
```

The robot will start moving it until no new frontiers are found. The frontiers will be published on ROS and the visualization enabled on rViz.

**Test Configuration**:

- [ ] Real robot
- [x] Gazebo simulation
- [ ] Other (please specify)

# Checklist:

- [x] My code follows the style guidelines of this project (Travis CI is passing)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by eborghi10** [comment]
The Linter is complaining! You can reproduce it locally with this command:

```bash
catkin lint -W2 --strict --explain /path/to/create_autonomy
```

**Comment by eborghi10** [comment]
Update the branch with latest `kinetic-devel` and you will need to use `RVIZ=true` instead of `RVIZ_CONFIG`.

File: docker/create_ros_melodic_gazebo9/Dockerfile [hunk 1]
```diff
@@ -24,3 +24,20 @@ RUN apt-get update && \
     xargs -a packages.txt apt-get install -y
 
 USER $USER
+# Install custom cartographer
+ENV CARTO_WS="/home/${USER}/catkin_ws"
+RUN mkdir ${CARTO_WS}
+WORKDIR ${CARTO_WS}
+RUN wstool init src && \
+    wstool merge -t src \
+        https://raw.githubusercontent.com/BarceloChristian/cartographer_ros/ros_maps/cartographer_ros.rosinstall && \
```
Comment (eborghi10) [inline docker/create_ros_melodic_gazebo9/Dockerfile]:
```suggestion
        https://raw.githubusercontent.com/RoboticaUtnFrba/cartographer_ros/ros_maps/cartographer_ros.rosinstall && \
```

File: docker/create_ros_melodic_gazebo9/Dockerfile [hunk 2]
```diff
@@ -24,3 +24,20 @@ RUN apt-get update && \
...
+    wstool update -t src
+RUN src/cartographer/scripts/install_proto3.sh --local
+RUN rosdep install --from-paths src --ignore-src --rosdistro=${ROS_DISTRO} -y -r --os=ubuntu:bionic
+# Change again to root in order to install in /opt
+USER root
+RUN /bin/bash -c "source /opt/ros/melodic/setup.bash && \
+    catkin_make_isolated --install --use-ninja --install-space /opt/ros/melodic"
```
Comment (eborghi10) [inline docker/create_ros_melodic_gazebo9/Dockerfile]:
```suggestion
RUN /bin/bash -c "source /opt/ros/${ROS_DISTRO}/setup.bash && \
    catkin_make_isolated --install --use-ninja --install-space /opt/ros/${ROS_DISTRO}"
```

File: navigation/mapping_exploration/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 3.0.2)
+project(mapping_exploration)
```
Comment (eborghi10) [inline navigation/mapping_exploration/CMakeLists.txt]: Just to be consistent with the rest of the packages:

```suggestion
project(ca_mapping_exploration)
```

File: navigation/mapping_exploration/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+
+find_package(catkin REQUIRED)
+
+catkin_package()
```
Comment (eborghi10) [inline navigation/mapping_exploration/CMakeLists.txt]: Install the launch directory.

File: navigation/mapping_exploration/launch/explore_lite.launch [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<launch>
+<node pkg="explore_lite" type="explore" respawn="true" name="explore" output="screen">
+  <param name="robot_base_frame" value="create1_tf/base_link"/>
```
Comment (eborghi10) [inline navigation/mapping_exploration/launch/explore_lite.launch]: As it's being done in order launch files, use arguments to get the robot ID and, from there, create the Tf prefix and namespace.

File: navigation/mapping_exploration/launch/explore_lite.launch [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<launch>
+<node pkg="explore_lite" type="explore" respawn="true" name="explore" output="screen">
```
Comment (eborghi10) [inline navigation/mapping_exploration/launch/explore_lite.launch]: The namespace should be set here and removed it from the topic parameter.

```suggestion
<node pkg="explore_lite" type="explore" respawn="true" name="explore" output="screen" ns="$(arg ns)">
```

File: navigation/mapping_exploration/launch/explore_lite.launch [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  <param name="robot_base_frame" value="create1_tf/base_link"/>
+  <param name="move_base_topic" value="create1/move_base"/>
+  <param name="costmap_topic" value="map"/>
+  <param name="costmap_updates_topic" value="map_updates"/>
```
Comment (eborghi10) [inline navigation/mapping_exploration/launch/explore_lite.launch]: Move all the constant parameters to a YAML file (as it's being done in `move_base` for example) inside a `config` directory.

File: remote.repos [hunk 8]
```diff
@@ -15,3 +15,7 @@ repositories:
     type: git
     url: https://github.com/RoboticaUtnFrba/viso2.git
     version: kinetic
+  explore_lite:
```
Comment (eborghi10) [inline remote.repos]: Move it alphabetically.

**Review by eborghi10 — CHANGES_REQUESTED** [review]
I made a few comments, please update it with latest `kinetic-devel` so I can give it a try.

File: docker/create_ros_melodic_gazebo9/Dockerfile (hunk 1 above)
Comment (BarceloChristian) [reply docker/create_ros_melodic_gazebo9/Dockerfile]: Thank you for catching this, fixed!

File: docker/create_ros_melodic_gazebo9/Dockerfile (hunk 2 above)
Comment (BarceloChristian) [reply docker/create_ros_melodic_gazebo9/Dockerfile]: Used `${ROS1_DISTRO}` as in the upper level layer

File: navigation/mapping_exploration/CMakeLists.txt (hunk 3 above)
Comment (BarceloChristian) [reply navigation/mapping_exploration/CMakeLists.txt]: Done!

File: navigation/mapping_exploration/CMakeLists.txt (hunk 4 above)
Comment (BarceloChristian) [reply navigation/mapping_exploration/CMakeLists.txt]: Done!

File: navigation/mapping_exploration/launch/explore_lite.launch (hunk 5 above)
Comment (BarceloChristian) [reply navigation/mapping_exploration/launch/explore_lite.launch]: Done too!

File: navigation/mapping_exploration/launch/explore_lite.launch (hunk 6 above)
Comment (BarceloChristian) [reply navigation/mapping_exploration/launch/explore_lite.launch]: Done!

File: navigation/mapping_exploration/launch/explore_lite.launch (hunk 7 above)
Comment (BarceloChristian) [reply navigation/mapping_exploration/launch/explore_lite.launch]: Done! Also wrote some explanation of the parameters in the Yaml

File: remote.repos (hunk 8 above)
Comment (BarceloChristian) [reply remote.repos]: That's called a `nit` hahaha, Done!

File: navigation/ca_mapping_exploration/package.xml [hunk 9]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>ca_mapping_exploration</name>
+  <version>0.0.0</version>
+  <description>Exploratory mapping software based on frontiers</description>
+  <maintainer email="christianbarcelo@ekumenlabs.com">Christian Barcelo</maintainer>
+  <license>BSD</license>
+
+  <buildtool_depend>catkin</buildtool_depend>
+
```
Comment (eborghi10) [inline navigation/ca_mapping_exploration/package.xml]: Dependencies here? m_explore, cartographer/any other mapping technique?

File: navigation/ca_mapping_exploration/launch/explore_lite.launch [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<launch>
+    <arg name="robot_id" default="$(optenv ID 1)" doc="Unique identifier of the robot [1-Inf.)"/>
+    <arg name="ns" default="create$(arg robot_id)"/>
+    <node pkg="explore_lite" type="explore" respawn="true" name="explore" output="screen" ns="$(arg ns)">
+        <rosparam command="load"
+               file="$(find ca_mapping_exploration)/config/explore_lite.yaml"/>
+    </node>
+</launch>
```
Comment (eborghi10) [inline navigation/ca_mapping_exploration/launch/explore_lite.launch]:
```suggestion
    <arg name="robot_id" default="$(optenv ID 1)"  doc="Unique identifier of the robot [1-Inf.)"/>
    <arg name="ns"  value="create$(arg robot_id)" doc="Namespace of the robot. By default: create1"/>
    
    <node pkg="explore_lite" type="explore" respawn="true" name="explore" output="screen" ns="$(arg ns)">
        <rosparam command="load" file="$(find ca_mapping_exploration)/config/explore_lite.yaml"/>
    </node>
</launch>

```

File: navigation/ca_mapping_exploration/config/explore_lite.yaml [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+costmap_updates_topic: "/map_updates"
+# Frontiers visualization
+visualize: true
+# Refresh frontier goals rate
+planner_frequency: 0.33
+# Timeout to consider a frontier unreachable
+progress_timeout: 30. #s
+# Scale to weigh frontiers based on distance
+potential_scale: 3.0
+orientation_scale: 0.0
```
Comment (eborghi10) [inline navigation/ca_mapping_exploration/config/explore_lite.yaml]: These seem to be interesting parameters.

File: docker/create_ros_melodic_gazebo9/Dockerfile [hunk 12]
```diff
@@ -24,3 +24,21 @@ RUN apt-get update && \
     xargs -a packages.txt apt-get install -y
 
 USER $USER
+# Install custom cartographer
```
Comment (eborghi10) [inline docker/create_ros_melodic_gazebo9/Dockerfile:28]:
```suggestion

# Install custom cartographer
```

**Review by eborghi10 — CHANGES_REQUESTED** [review]
A few silly requests. I'll try it in a moment.

File: remote.repos [hunk 13]
```diff
@@ -1,4 +1,8 @@
 repositories:
+  explore_lite:
+    type: git
+    url: https://github.com/RoboticaUtnFrba/m-explore.git
+    version: master
```
Comment (eborghi10) [inline remote.repos]: Move this to `rpi.repos` please.

**Review by eborghi10 — CHANGES_REQUESTED** [review]

**Comment by eborghi10** [comment]
@BarceloChristian can you add the option in RViz to visualize the frontiers?

**Review by eborghi10 — APPROVED** [review]
Awesome feature!

Since this PR is not related to the issue that I'm having, I''ll approve it!

**Pull request merged**